src/librustc_resolve/lib.rs

```diff
@@ -4304,7 +4305,15 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                     let mut path_segments = path_segments.clone();
                     path_segments.push(ast::PathSegment::from_ident(ident));
 
-                    if !in_module_is_extern || name_binding.vis == ty::Visibility::Public {
+                    let is_extern_crate_that_also_appears_in_prelude =
+                        name_binding.is_extern_crate() &&
+                        self.session.rust_2018() &&
+                        self.extern_prelude.contains(&ident.name);
```

Comment:
if my new understanding is correct, I think we can just remove this `self.extern_prelude.contains(&ident.name)` line and we don't need to special case std at all.